src/hotspot/share/nmt/memLogRecorder.hpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_NMT_MEMLOGRECORDER_HPP
+#define SHARE_NMT_MEMLOGRECORDER_HPP
+
+#include "memory/allocation.hpp"
+#include "nmt/nmtCommon.hpp"
+#include "runtime/globals.hpp"
+
+class NMT_LogRecorder : public StackObj {
+#if defined(LINUX) || defined(__APPLE__)
+  pthread_mutex_t _mutex;
+#elif defined(WINDOWS)
+ // ???
+#endif
```

Comment:
You need to use existing synchronization API here. Either Mutex or PlatformMutex depending on initialization constraints; maybe a Semaphore if the others can't work.

Reply:
We have synchronization abstractions so that we don't litter shared code with platform specific code - and don't yu need a solution for Windows? We also need to ensure synchronization plays nicely with safepoints and yes to watch for potential deadlock issues. Testing can never prove the absence of deadlock.